### PR TITLE
APPS-1333 Handle namespace backup retries individually

### PR DESCRIPTION
**Pull request by korotkov-aerospike**

### CODE REVIEW

File: pkg/service/backup_routine_handler.go [hunk 1]
```diff
@@ -140,64 +159,74 @@ func (h *BackupRoutineHandler) runFullBackupInternal(ctx context.Context, now ti
...
-	namespaces, err := getNamespacesToBackup(h.namespaces, client.AerospikeClient())
-	if err != nil {
-		return err
-	}
-
-	for _, namespace := range namespaces {
-		backupFolder := getFullPath(h.routineName, h.backupFullPolicy, namespace, upperBound)
-		handler, err := h.backupService.BackupRun(ctx, h.backupRoutine, h.backupFullPolicy, client,
-			h.storage, h.secretAgent, timebounds, namespace, backupFolder)
+func (h *BackupRoutineHandler) prepareCluster() (*backup.Client, []string, error) {
```
Comment (reugn) [inline pkg/service/backup_routine_handler.go]: Consider renaming to `prepareClusterWithRetries` for better readability.

File: pkg/service/backup_routine_handler.go [hunk 2]
```diff
@@ -140,64 +159,74 @@ func (h *BackupRoutineHandler) runFullBackupInternal(ctx context.Context, now ti
...
+	var client *backup.Client
+	var namespaces []string
```
Comment (reugn) [inline pkg/service/backup_routine_handler.go]: Can be grouped.

File: pkg/service/backup_routine_handler.go [hunk 3]
```diff
@@ -96,37 +124,28 @@ func getNamespacesToBackup(namespaces []string, client backup.AerospikeClient) (
...
 		h.logger.Info("Full backup is currently in progress, skipping full backup")
-		return nil
+		return
 	}
 
-	client, err := h.clientManager.GetClient(h.backupRoutine.SourceCluster)
+	client, namespaces, err := h.prepareCluster()
 	if err != nil {
-		return err
+		return
```
Comment (reugn) [inline pkg/service/backup_routine_handler.go:134]: Do we log the error somewhere?

File: pkg/service/backup_routine_handler.go [hunk 4]
```diff
@@ -96,37 +124,28 @@ func getNamespacesToBackup(namespaces []string, client backup.AerospikeClient) (
...
-		return err
+	for _, namespace := range namespaces {
+		h.fullBackupHandlers[namespace] = h.startNamespaceBackup(ctx, namespace, now, client)
 	}
 
-	err = h.waitForFullBackups(ctx, now)
+	err = h.waitForFullBackups(ctx)
 	if err != nil {
-		return err
+		return
```
Comment (reugn) [inline pkg/service/backup_routine_handler.go:148]: Do we log the error somewhere?

File: pkg/service/retryable_backup_handler.go [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/aerospike/backup-go/models"
+)
+
+type RetryableBackupHandler struct {
```
Comment (reugn) [inline pkg/service/retryable_backup_handler.go]: Unexport the type.

File: pkg/service/retryable_backup_handler.go [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+	sync.RWMutex
+	handler BackupHandler
+	errCh   chan error
+}
+
```
Comment (reugn) [inline pkg/service/retryable_backup_handler.go:16]: Add the interface validation variable.

File: pkg/service/retryable_backup_handler.go [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+func newRetryableBackupHandler(
+	ctx context.Context,
+	retry *RetryService,
+	start func(ctx context.Context) (BackupHandler, error),
+	onFail func(ctx context.Context),
+	onSuccess func(ctx context.Context, stats *models.BackupStats) error,
+) *RetryableBackupHandler {
```
Comment (reugn) [inline pkg/service/retryable_backup_handler.go]: Return `BackupHandler`.

File: pkg/service/retryable_backup_handler.go [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+			return fmt.Errorf("backup failed: %w", err)
+		}
+
+		return onSuccess(ctx, handler.GetStats())
+	}
+
+	// Start the backup process with retries
+	go func() {
+		h.errCh <- retry.retry("backup", processBackup)
+	}()
```
Comment (reugn) [inline pkg/service/retryable_backup_handler.go:64]: This routine initialization is unexpected in the constructor.

File: pkg/service/backup_routine_handler.go [hunk 9]
```diff
@@ -140,64 +159,74 @@ func (h *BackupRoutineHandler) runFullBackupInternal(ctx context.Context, now ti
...
-	for namespace, handler := range h.fullBackupHandlers {
-		backupFolder := getFullPath(h.routineName, h.backupFullPolicy, namespace, backupTimestamp)
-		err := handler.Wait(ctx)
-		if err != nil {
-			backupFailureCounter.Inc()
-			h.logger.Info("Delete failed backup folder",
-				slog.String("path", backupFolder),
+func (h *BackupRoutineHandler) startNamespaceBackup(
+	ctx context.Context, namespace string, now time.Time, client *backup.Client,
+) *RetryableBackupHandler {
```
Comment (reugn) [inline pkg/service/backup_routine_handler.go]: Return `BackupHandler`.

File: pkg/service/backup_routine_handler.go [hunk 10]
```diff
@@ -140,64 +159,74 @@ func (h *BackupRoutineHandler) runFullBackupInternal(ctx context.Context, now ti
...
+			aggregatedErr = errors.Join(aggregatedErr, err)
 		}
 	}
-	duration := float64(time.Since(startTime).Milliseconds())
-	h.logger.Debug("Finished full backup", slog.Float64("duration_ms", duration))
-	backupDurationGauge.Set(duration)
-	return nil
+
+	durationMs := float64(time.Since(startTime).Milliseconds())
+	h.logger.Debug("Finished full backup", slog.Float64("duration_ms", durationMs))
```
Comment (reugn) [inline pkg/service/backup_routine_handler.go]: The log message should change if any error was returned.

File: pkg/service/backup_routine_handler.go [hunk 11]
```diff
@@ -140,64 +159,74 @@ func (h *BackupRoutineHandler) runFullBackupInternal(ctx context.Context, now ti
...
+	}
+	return model.TimeBounds{}
+}
+
+func (h *BackupRoutineHandler) waitForFullBackups(ctx context.Context) error {
+	startTime := time.Now()
+
+	var aggregatedErr error
+	for _, handler := range h.fullBackupHandlers {
+		if err := handler.Wait(ctx); err != nil {
```
Comment (reugn) [inline pkg/service/backup_routine_handler.go:222]: Is it acceptable for the first handler to potentially block the others due to sequential execution?

**Review by reugn — COMMENTED** [review]
Add an example for partial state where the backup of one of the namespaces failed.

File: pkg/service/backup_routine_handler.go (hunk 11 above)
Comment (korotkov-aerospike) [reply pkg/service/backup_routine_handler.go:222]: no, all process are working in separate gorouines

File: pkg/service/retryable_backup_handler.go (hunk 8 above)
Comment (korotkov-aerospike) [reply pkg/service/retryable_backup_handler.go:64]: agree, renamed the method, its not constructor anymore

File: pkg/service/backup_routine_handler.go (hunk 4 above)
Comment (korotkov-aerospike) [reply pkg/service/backup_routine_handler.go:148]: in retry service, it logs all failures.

File: pkg/service/backup_routine_handler.go (hunk 3 above)
Comment (korotkov-aerospike) [reply pkg/service/backup_routine_handler.go:134]: in retry service, it logs all failures.

**Review by reugn — APPROVED** [review]

**Pull request merged**